### DISCUSSION TITLE
Bug: Solver Status is ABNORMAL

### DISCUSSION THREAD

**Discussion by ShivKJ** [google/or-tools#3649] [Linear Solver questions]
**What version of OR-Tools and what language are you using?**
Version: 9.4.1874

Language: Python (3.10.4 (v3.10.4:9d38120e33, Mar 23 2022, 17:29:05) [Clang 13.0.0 (clang-1300.0.29.30)])

**Which solver are you using (e.g. CP-SAT, Routing Solver, GLOP, BOP, Gurobi)**
GLOP

**What operating system (Linux, Windows, ...) and version?**
macOS-12.6.1-arm64-arm-64bit

**What did you do?**
I created a linear programming model and used `GLOP_LINEAR_PROGRAMMING` to solve this. I am getting `ABNORMAL` as solver status. When I used other solver such as `CPLEX_LINEAR_PROGRAMMING` I am getting solution status as `OPTIMAL`.

I tried switching off some of the constraint to find out which constraint could have caused the issue. This constraint which caused the issue, has the following form,

```
let s, d1, d2 be NumVar with same lower bound as 0 but with different upper bound.

solver.Add(s == sum([d1 , d2])) 
```

Removing this (equality) constraint gives `OPTIMAL` solution. 

1. Is it possible that some solver type can give `OPTIMAL` solution while other `ABNORMAL`?
2. Is there a problem with equality constraint when using `GLOP_LINEAR_PROGRAMMING`?

**Comment by Regista6** [comment]
Can you try enabling dual simplex?. I faced this with GLOP once, and enabling dual simplex solved this.

```python
pywraplp.Solver.SetSolverSpecificParametersAsString(solver, "use_dual_simplex:1")
```

**Comment by ShivKJ** [comment]
@Regista6 I am again getting the same issue.

Here is how I have set the parameter,

```python
solver = Solver('LP-Problem', Solver.GLOP_LINEAR_PROGRAMMING)
solver.SetSolverSpecificParametersAsString("use_dual_simplex:1") # this returns True btw
```

Here is the what gets printed on terminal (showing only one line, other lines repeat this),
```
E0123 14:55:20.807219     1 linear_solver.cc:1879] No solution exists. MPSolverInterface::result_status_ = MPSOLVER_ABNORMAL
```

**Comment by lperron** [comment]
Glop is more pedantic w.r.t. numerical errors.
There should be something in the log that tells you which tolerance is violated.

You can increase the tolerance afterwards to get the optimal status, or disable solution checking.

**Reply by ShivKJ** [reply]
@lperron whenever I am calling `solution_value()` for a given variable, above line gets printed on terminal.

>You can increase the tolerance afterwards to get the optimal status, or disable solution checking.

Are you referring `solver.VerifySolution` function for it disabling/increasing tolerance?

I used `solver.VerifySolution(-1, False)` to disable solution checking (In function doc, it is mentioned: `If "tolerance" is negative, it will be set to infinity()`.). But still getting the same error.

How to disable solution checking/increase tolerance?

**Comment by lperron** [comment]
There are parameters:

https://github.com/google/or-tools/blob/8179abf8346b02c9aa39d356177fdbd6315f2c34/ortools/glop/parameters.proto#L136

and

https://github.com/google/or-tools/blob/8179abf8346b02c9aa39d356177fdbd6315f2c34/ortools/glop/parameters.proto#L148

Solution verification can only be controlled in C++.
Laurent Perron | Operations Research | ***@***.*** | (33) 1 42 68 53
00

**Comment by ShivKJ** [comment]
I was looking at `GlopParameters` class in python. Can it be used to inform solver for `GLOP` parameters.

```python
from ortools.glop.parameters_pb2 import GlopParameters

p = GlopParameters(primal_feasibility_tolerance=100000))
```

1. What are the other ways to pass this info to solve? 
2. Are you planning to expose these setting in python?

**Comment by lperron** [comment]
Use the string version and
linear_solver.SetSolverSpecificParametersAsString('primal_feasibility_tolerance:100000')

which seems really big to me :-)
Laurent Perron | Operations Research | ***@***.*** | (33) 1 42 68 53
00

**Reply by ShivKJ** [reply]
@lperron I wanted to test the idea of disabling the check, hence passed a very high value.

**Comment by lperron** [comment]
ok
Laurent Perron | Operations Research | ***@***.*** | (33) 1 42 68 53
00

**Comment by ShivKJ** [comment]
@lperron 

Let me know if the below solution will work,

1. Install all the prerequisite for [ortools](https://developers.google.com/optimization/install/python/source_linux#prerequisites)
2. `git clone https://github.com/google/or-tools`
3. `git fetch --all --tags --prune`
4. modify the [primal_feasibility_tolerance](https://github.com/google/or-tools/blob/8179abf8346b02c9aa39d356177fdbd6315f2c34/ortools/glop/parameters.proto#L136) and [dual_feasibility_tolerance](https://github.com/google/or-tools/blob/8179abf8346b02c9aa39d356177fdbd6315f2c34/ortools/glop/parameters.proto#L148) values
5.  `make third_party`
6. `make python`
7. `pip install build_make/python`

**Comment by lperron** [comment]
No need for that, and make is deprecated.

Just change the parameter on the linear solver before solving.

**Reply by ShivKJ** [reply]
You meant using this `solver.SetSolverSpecificParametersAsString('primal_feasibility_tolerance:VALUE')`?

I am still getting the error (adding one line of what I got on terminal),

```
E0124 11:52:09.572049     1 linear_solver.cc:1879] No solution exists. MPSolverInterface::result_status_ = MPSOLVER_ABNORMAL
```

**Comment by lperron** [comment]
Can you do

model_str = solver.ExportModelAsMpsFormat(False, True)

Save it to file, and send me the file ?
Laurent Perron | Operations Research | ***@***.*** | (33) 1 42 68 53
00

**Reply by ShivKJ** [reply]
@lperron 

Thanks for extending for help. I have been exploring [GLOP parameters](https://github.com/google/or-tools/blob/8179abf8346b02c9aa39d356177fdbd6315f2c34/ortools/glop/parameters.proto) and able to resolve this by changing value of `solution_feasibility_tolerance` to a higher value (default value is 1e-6). Have tested it on multiple inputs data.

Correct me if I am wrong, these kind of parameters (such as `solution_feasibility_tolerance, primal_feasibility_tolerance, dual_feasibility_tolerance etc`) do not affect the LP (feasible/optimal) solution, they are used for validations.

**Reply by lperron** [reply]
yes
